crates/matrix-sdk-common/src/lib.rs

```diff
@@ -10,17 +10,13 @@
 
 pub use async_trait::async_trait;
 pub use instant;
+pub use ruma;
```

Comment:
I definitely think the current re-export has some value even though I expect myself to stop using it in my one SDK-dependent project once I no longer use it as a git dep.

For a re-export from `-common` I'm definitely wondering what the benefit would be.

Reply:
My main take on re-exports is, that they must be considered _if_ the API of the given crate export them directly (either as param or return value), one must consider exporting the given type directly for ease of use from the lowest point. The background is that if you have multiple dependencies in your try depending on different (rather than a common, re-exported) versions of the same dependency the highest level will end up in an unresolvable incompatible state. This is particularly annoying when mixing git and release-versions.

Example:

```
Client
 +- agent-smith
 |  +- ruma 0.4
 |  +- matrix-common
 |
 +- matrix-sdk
 |  +- matrix-crypto
 |  |  +- ruma 0.5
 |  +- matrix-common
```

Now any `ruma`-type returned by `agent-smith` will be incompatible with the `matrix-sdk` input `ruma`-types - even if they aren't any different at all. This leads to a rust build error on incompatible types, that can't be resolved by `Client` but only by agent-smith releasing another version of the library now depending on `ruma 0.5`. 

If you are leaking types from a dependency through your API, you _are_ already exporting them (and they are often forgotten or ignored when it comes to SemVer). Not re-exporting the crate as a whole just makes it a lot harder for someone to build upon them and keep things up to date in the longer-term, especially once other dependencies are involved.

Through the lowest API exporting the used version, everything building upon can just build upon the export and thus be sure to always be compatible, rather than looking up the version used in the dependency and adding it to their own Cargo.toml. Arguably, whenever that lowest crate changes the dependency in a non-semver-compatible way, it must release a new major version, too, to avoid the problem mentioned above (which is rarely done properly). The alternative is to not export types of dependencies, but that seems inconvenient for us for now.

I know re-exports are controversial, but I wanted to explain the reasoning behind it and why I think in particular ruma makes a lot of sense for us to export (we depend highly on it for various types).

Reply:
By this argument, every library should re-exports every public dependency. This is not at all how things work in the overall Rust ecosystem and I think the Matrix SDK shouldn't stray too far from that.

I just don't think that the problem being solved is big enough to go for such an "ugly" and incomplete solution¹, with the error messages for that kind of type mismatch being pretty good these days and Cargo has support for explicitly marking dependencies as public or private which should help as well (though there doesn't seem to be any recent movement towards stabilization unfortunately – https://github.com/rust-lang/cargo/issues/6129).

¹ Why I say that is:

* I like to nest my imports and with re-exports you get extra unnecessary rightwards drift there
* Boundaries between crates / projects can get less clear
* It's also unclear where you import things from when multiple crates re-export them, a real-world example I've come across is `reqwest`'s re-exports of `http` crate types and the whole-crate `axum::http` re-export (though I was actually in favor of that particular one being added)
* You can still have multiple dependencies to different versions of one lower-level crate

Reply:
> By this argument, every library should re-exports every public dependency.

Is they are fundamental, in that they themselves export them via their API, yes, probably. The argument still holds and I don't think you pointing out its final consequence changes much about that.

>This is not at all how things work in the overall Rust ecosystem and I think the Matrix SDK shouldn't stray too far from that.

Just pointing out that below you are naming the popular `reqwest`-library as an example doing it and I'd also add the [widely popular `parking_lot`, exporting the lower-level `lock_api`](https://docs.rs/parking_lot/latest/parking_lot/index.html) to that list. So I reject the premise that "this is not how things work in overall Rust ecosystem" - and even if, this might be because this problem hasn't been run into as much yet. "It is not being done much yet" is, by itself, not really an argument...

> I like to nest my imports and with re-exports you get extra unnecessary rightwards drift there

What do you mean with `rightwards drift`?

> Boundaries between crates / projects can get less clear

Arguably, they already are because the API is exporting _types external to its crate_ - which makes navigating the docs pretty annoying, btw, as you can't click half the params and return types but instead have to navigate to the ruma-api yourself every time, ensure you are looking at the exact same version and find the type manually. I am not sure if the re-export fixes that (especially for higher level APIs, like if matrix-sdk used a ruma-type), but it seems to do so on `parking_lot`. 

> You can still have multiple dependencies to different versions of one lower-level crate 

You mean _despite_ re-exporting, you can still have different versions? Yeah, sure, that's true either way, and thus not really an argument against it though. Never said, it would solve that issue.

> It's also unclear where you import things from when multiple crates re-export them, a real-world example I've come across is reqwest's re-exports of http crate types

That's the only argument, why I _usually_ don't like re-exporting and abstain from it as well. It makes things a bit more vague... but in practice this leads only to a problem if different things re-export the same thing - and that would cause the same issue of incompatible ruma types. One must be aware of that, but that's differen from what I am proposing here: In what I am proposing only common would re-export ruma, the others won't (but refer to common to be the place it comes from) and we'd generally recommend using `common::ruma` as the entry point when using the lib. In practice any _other_ import of `ruma` won't be compatible with `matrix-sdk` anyways, so I don't see the point of making people jump through the hoops of maintaining their own Crates.toml dependency.

> with the error messages for that kind of type mismatch being pretty good these days and Cargo has support for explicitly marking dependencies as public or private which should help as well (though there doesn't seem to be any recent movement towards stabilization unfortunately – https://github.com/rust-lang/cargo/issues/6129).

Agreed, it would be nice (though not fixing the problem entirely), if this was merged. But it hasn't moved in over two years so ... not really an argument to wait for that instead.

> "ugly" and incomplete solution

This is what the argument boils down at the end though, that you/some consider this _ugly_ on principle - and I agree it doesn't feel right to do that generally. But especially considering there isn't any _complete_ solution (as of now, nor on the horizon) and other the icky-ness there isn't much of any drawbacks (I can see), I'd rather take a practical approach and that is: one way is a lot easier for the users of the API and one is a lot harder. I'd rather take the easier one, even if it feels likely ugly or icky, if that doesn't come with any clear drawbacks. And I don't see that this does.

Reply:
Btw, not a hill I am willing to die on. I care 4/10 about this, just made it for convenience and am fine to change it back, but don't have seen any good argument why that would be better.

Reply:
> Just pointing out that below you are naming the popular reqwest-library as an example doing it

Actually, reqwest doesn't do a whole-crate re-export like here but I guess it doesn't matter that much.

Honestly overall it sounds like we're almost on a similar page, as I said I see some value in the existing re-export. I just saw `-common` as more of an implementation detail and not something users would interact with directly, so if `ruma` then becomes a `pub use matrix_sdk_common::ruma;` re-export in the docs rather than just `pub use ruma;`, that seems like it would just add confusion and not really help users in any way.

Sorry if my previous reply was a bit intense, I really don't want to die on that hill either ;)